src/pages/RFOX/hooks/useEpochHistoryQuery.ts

```diff
@@ -1,38 +1,93 @@
+import type { QueryClient } from '@tanstack/react-query'
+import { useQueryClient } from '@tanstack/react-query'
+import dayjs from 'dayjs'
 import { useMemo } from 'react'
+import { arbitrum } from 'viem/chains'
 import { useQuery } from 'wagmi/query'
+import { viemClientByNetworkId } from 'lib/viem-client'
 
 import type { EpochMetadata } from '../types'
+import { affiliateRevenueQueryFn, getAffiliateRevenueQueryKey } from './useAffiliateRevenueQuery'
+import {
+  blockNumberByTimestampQueryFn,
+  getBlockNumberByTimestampQueryKey,
+} from './useBlockNumberByTimestampQuery'
 
 type EpochHistoryQueryKey = ['epochHistory']
 
+const RFOX_FIRST_EPOCH_START_TIMESTAMP = BigInt(dayjs('2024-07-01T00:00:00Z').unix())
+
+const client = viemClientByNetworkId[arbitrum.id]
+
+// The query key excludes the current timestamp so we don't inadvertently end up with stupid things like reactively fetching every second etc.
+// Instead we will rely on staleTime to refetch at a sensible interval.
 export const getEpochHistoryQueryKey = (): EpochHistoryQueryKey => ['epochHistory']
 
 export const getEpochHistoryQueryFn =
-  () =>
-  (_: { queryKey: EpochHistoryQueryKey }): { data: EpochMetadata[] } => {
-    // TODO: fetch the epoch info from our indexer
-    return {
-      data: [
-        {
-          startBlockNumber: 0n,
-          endBlockNumber: 0n,
-          startTimestamp: 0n,
-          endTimestamp: 0n,
-          distributionAmountRuneBaseUnit: 1234567890n,
-        },
-      ],
+  (queryClient: QueryClient) => async (): Promise<EpochMetadata[]> => {
+    const now = dayjs().unix()
+    let startTimestamp = RFOX_FIRST_EPOCH_START_TIMESTAMP
+
+    // using queryClient.fetchQuery here is ok because block timestamps do not change so reactivity is not needed
+    let startBlockNumber = await queryClient.fetchQuery({
+      queryKey: getBlockNumberByTimestampQueryKey(startTimestamp),
+      queryFn: blockNumberByTimestampQueryFn,
+    })
+
+    const epochHistory = []
+
+    while (startTimestamp < now) {
```

Comment:
q: Can we be rate-limited/exhaust caps in etherscan API resulting in failures?

If so, it may make sense to rely on block as a criteria vs. timestamps (i.e `while (startBlockNumber.timestamp < now)`

Reply:
`startBlockNumber` is a bigint, but I get what you're saying - I'll rector to use viem client to do a search similar to other parts of the app

Reply:
will do in follow-up to keep things moving